Use client interceptors added in Ratpack 1.6

llinder opened the pull request.

llinder committed:
Use client interceptors added in Ratpack 1.6

llinder commented on src/main/java/ratpack/zipkin/ServerTracingModule.java:
```diff
@@ -50,11 +51,27 @@ protected void configure() {
+        getProvider(ClientTracingInterceptor.class);
+
+    Provider<HttpClient> httpClientProvider = () ->
+        Exceptions.uncheck(() -> HttpClient.of(spec -> {
```
As originally pointed out by @beckje01 `This isn't perfect, I think it needs to be additive right now meaning it can't take over the HttpClient directly yet without ratpack changes...`

I'm still unsure how to address this.  As far as I can tell there is no way to make this additive but if anyone has ideas I'm open to trying other approaches.

beckje01 replied:
I don't think we can do any better right now

jeff-blaisdell replied:
I think I'm missing why the previous pattern of instantiating a second client, and have it only inject when the @Zipkin annotation exists doesn't work?

jeff-blaisdell replied:
Personally, I think I like it additive better as it allows me greater control over what I choose to trace.  Additionally, if there was a bug in this library - theoretically I could choose to remove the annotation and use the default client with no changes.

llinder replied:
Thanks @jeff-blaisdell.  Maybe we should raise an issue against Ratpack to introduce an API which takes an existing HttpClient that we can extend with additional config.

llinder replied:
@jeff-blaisdell would it be enough to _not_ replace the default HttpClient instance and only add the annotated version like before?  This doesn't solve the additive config problem but it would allow you to opt out of using the instrumented client.

jeff-blaisdell replied:
I just wasn't understanding the problem.  To paraphrase:
The intention was it would be nice if somehow we could extend default client configuration with additional configs that are additive between the 2 clients rather than independent.  Agreed, that would be nice.

My point above is tangential to the config issue in that my preference is to have 2 independent client that I has the consumer of the library can choose between; rather than opt'ing me fully into Zipkin's variant.

janderson007 replied:
Does the http client instance need to be provided directly by the module?  Why not just provide the interceptors via the module and ask the application to register them with their http client?

llinder replied:
Think I have solved both concerns.  First I removed unannotated binding for `HttpClient` and second I changed the instrumented HttpClient provider to use `copyWith` to make the instrumentation additive to the default client as apposed to creating a new client.  This should allow modules configured before this one to set `HttpClient` configuration and this module will inherit the config.

llinder replied:
@janderson007 that would work as well however right now users expect that they can just use `@Zipkin` annotation to get an instrumented client.  Removing this means services would have to do more work and we would have to document this well.  Also in practice it was hard enough to communicate that you needed `@Zipkin` so I suspect that requiring more steps would bring even more questions when things don't work as expected.

drmaas replied:
@llinder another point against the `@Zipkin` annotation as implemented is that it makes it impossible to disable zipkin by selectively installing/choosing not to install the zipkin module. We recently wanted to disable zipkin and had to make a code change to comment out the annotation. 
This might be a side PR but it would be nice to provide a config to the module to enable/disable tracing without code changes.

llinder replied:
@drmaas thats a great suggestion.  Adding some configuration to disable client instrumentation should be an easy add.  To solve the use case of being able to remove the Zipkin Module without requiring code changes, I think we would have to deprecate the use of `@Zipkin` annotation and replace the default HttpClient binding with the instrumented one.  In the end of the day, if someone wants both an instrumented and un-instrumented instance they could bind a new un-instrumented instance with their own annotation.

For now I will leave this PR as is and write up a ticket for your configuration suggestion.  Once this is merged it should be fairly painless to add the config change and cut a new release.  For now I still have to get the last test to pass so I will keep fiddling with that.

llinder commented:
Looks like Travis doesn't appreciate this test https://github.com/openzipkin-contrib/brave-ratpack/blob/master/src/test/java/brave/http/ITZipkinHttpClientImpl.java#L105 . Locally all is fine but I guess Travis CPU restrictions is causing a deadlock :(

jeff-blaisdell commented on src/test/java/brave/http/ITZipkinHttpClientImpl.java:
```diff
@@ -63,6 +70,9 @@ void harnessSetup(Execution execution) {
         harness.yield(e -> client.get(URI.create(url(pathIncludingQuery)))).getValueOrThrow();
+        // Small delay to ensure the response interceptor is invoked before this method exits.
+        // This is needed for inherited test `redirect`
+        Thread.sleep(1);
```
nit-picky but I generally find leveraging [Spock Polling or Blocking Conditions](http://spockframework.org/spock/javadoc/1.0/spock/util/concurrent/AsyncConditions.html) makes tests more robust.  Though sometimes that takes some gymnastics to make happen.

llinder replied:
That makes a lot of sense.  These test use existing integration tests from the Brave project so hands are tied a bit.

jeff-blaisdell reviewed (commented):
Looks good to me!

llinder committed:
Changed client interceptor to handle multiple spans in the execution

While testing it appearts that the same execution is shared between multiple requests.  Hopefully this isn't the case
in production but if it is I don't understand how we could ever use the interceptor as instrumentation for tracing.

llinder commented:
> good work!
> 
> I think by default the IT doesnt expect implicit redirect handling. i guess we can add a test here for that or a TODO

Testing for implicit redirect handling would be useful.  I could overwrite that one test to do this as well I suppose.  Still I feel like there isn't much in the default integration tests that is getting in the way.  Mainly I'm struggling with making the Ratpack HTTP Client Interceptor APIs work, mostly because there isn't a way to pass state between the request and response interceptor without putting the state in the Execution context.  While using the Execution context _should_ be fine, it doesn't behave as expected with the test harness setup.  During tests the Execution context is being shared between requests which to my understanding should *not* be the case.

I think for now my game plan is to test this in a real application to see how it behaves.  If a simple real world app doesn't exhibit anything weird, then I can move on to sorting out the Travis CI difficulties :/

codefromthecrypt commented:
> Testing for implicit redirect handling would be useful. I could overwrite that one test to do this as well I suppose. Still I feel like there isn't much in the default integration tests that is getting in the way. Mainly I'm struggling with making the Ratpack HTTP Client Interceptor APIs work, mostly because there isn't a way to pass state between the request and response interceptor without putting the state in the Execution context. While using the Execution context should be fine, it doesn't behave as expected with the test harness setup. During tests the Execution context is being shared between requests which to my understanding should not be the case.

i see, so Execution context isn't request scoped. yeah that could be a
problem. A bounded weak hash map might be an ok alternative, if there
is no other bucket to use.

> I think for now my game plan is to test this in a real application to see how it behaves. If a simple real world app doesn't exhibit anything weird, then I can move on to sorting out the Travis CI difficulties :/

good idea to conserve energies! Thanks for the thorough reply!

drmaas commented:
> While using the Execution context should be fine, it doesn't behave as expected with the test harness setup

My understanding is that an `Execution` is not shared. They end up in a queue and get executed on various even loop threads.

In the `ExecHarness` class - it provides a single execution (by creating a new event loop and forking an execution on it) on which to perform activities.

I am interested to know what exactly happens to the `Execution` variables between request reception and response processing, and why things aren't working as expected in `ExecHarness`.

llinder committed:
Client tracing updates

- switched to a supplier pattern for acquiring the current execution
- removed redirect handling logic as it proved to be too hard to implement with the interceptor logic
- ignored upstream redirect IT test
- added special case logic for handling a non existant execution context which happens on redirects
- added special case context handling to ensure that removing a span from the registry doesn't break some tests.

llinder committed:
Ignoring problematic test on CI server

llinder commented:
> My understanding is that an `Execution` is not shared. They end up in a queue and get executed on various even loop threads.

This ended up being correct, thanks for nudging me in the right direction!  I think with all of the ExecHarness issues I was confusing integration test issues and completely different runtime issues.

> In the `ExecHarness` class - it provides a single execution (by creating a new event loop and forking an execution on it) on which to perform activities.

This statement does seem to match my experience in that every call to `harness.yield` was producing a shared execution for every HttpClient request.  In the end I worked around it by putting any `ClientSpanHolders` back into the registry.  This isn't ideal but without it the tests would fail.

> I am interested to know what exactly happens to the `Execution` variables between request reception and response processing, and why things aren't working as expected in `ExecHarness`.

When multiple requests are issued during a single test, each call to the interceptor request adds a new Span to the registry.  The behavior I see is that the previous Span is already in the registry when the second call happens.  Then later when the interceptor receive method is invoked we get the Span, which may or may not be the last one, and remove it from the registry.  Remove also happens to clear out all values queued up in the registry that match the type.  This is a different nit pick entirely but there isn't really a way to only take a single value without taking out all values and inserting things back.

When I test this with a standalone test app the Execution is __not__ shared which is what I would expect.  Ironically though, redirects invoke the interceptor receive without an Execution entirely at runtime.  End result is that redirects don't play well with the interceptor so I ended up having to just ignore them.  Even trying to do something like annotate the Span on redirect didn't work since the interceptor receive method is always called _before_ `onRedirect` so the Span is completed before the annotation can be applied.

And last but not least, the client interceptor receive method invocation is delayed.  This is usually a couple milliseconds.  In the screenshot the first Span is the server request/response.  Second span is a location span around the client call.  Third span is a remote http call generated by the client interceptor.

![image](https://user-images.githubusercontent.com/99359/61757132-e45b7c80-ad84-11e9-9c56-801ed7babc39.png)

Here is another example, this time using a ParallelBatch to perform several calls in parallel.

![image](https://user-images.githubusercontent.com/99359/61757241-6186f180-ad85-11e9-9b0a-6b022356c95e.png)

llinder commented on src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java:
```diff
@@ -112,10 +116,15 @@ private TraceContextHolder(final TraceContext context) {
         .ifPresent(execution::add);
+
+      maybeParent
+          .flatMap(parent -> parent.maybeGet(DefaultClientTracingInterceptor.ClientSpanHolder.class))
```
Copying the `ClientSpanHolder` forward on forked executions was the key to making the client interceptor work.  Effectively the client interceptor request and receive have different executions.  Fortunately the receive is a child of the request at runtime and copying forward the `ClientSpanHolder` saved the day.

codefromthecrypt replied:
add comment so tomorrow is saved, too? :D

llinder committed:
Leave comment on the ExecInitializer about the Client Span handling.

kirenjolly commented:
@llinder Is there any estimated time on when this PR will get merged and released, since we are facing a production issue due to context memory leak when using streaming requests.
It will be highly appreciated if we could get this fix soon :)

llinder commented:
@kirenjolly I think the only thing remaining is to pull the trigger and accept the differences between the interceptor approach vs. the older wrapped http client approach.  After getting the interceptor approach working its clear that both have their own pros/cons.  Reviving the wrapped client approach as a separate library for 1.6 and 1.7 is always another option so I don't see this MR as a one way street which reduces risk I suppose.

Since `requestStream` has been an issue for us as well I'm going to do a bit more testing with that.  In theory this should resolve the memory leak but I would like to see it run for a while and monitor heap usage at least.
